File: src/main/scala/org/joshjoyce/lumivore/sync/GlacierUploader.scala
package org.joshjoyce.lumivore.sync

import java.io.File
import java.util.concurrent.Executors

import com.amazonaws.auth.PropertiesCredentials
import com.amazonaws.event.{ProgressEvent, ProgressEventType, ProgressListener}
import com.amazonaws.services.glacier.AmazonGlacierClient
import com.amazonaws.services.glacier.transfer.{ArchiveTransferManager, UploadResult}
import org.jetlang.channels.Channel
import org.joshjoyce.lumivore.io.FileUtils
import org.joshjoyce.lumivore.util.LumivoreLogging

sealed trait GlacierUploadAttempt

case class CompleteUpload(filePath: String, vaultName: String, archiveId: String, percent: Int) extends GlacierUploadAttempt

case class PartialUpload(filePath: String, percent: Int) extends GlacierUploadAttempt

case class FailedUpload(filePath: String, vaultName: String, e: Throwable, percent: Int) extends GlacierUploadAttempt

case object Done extends GlacierUploadAttempt

class GlacierUploader(output: Channel[GlacierUploadAttempt]) extends LumivoreLogging {
  private var credentials: PropertiesCredentials = _
  private var client: AmazonGlacierClient = _
  private var atm: ArchiveTransferManager = _
  private val pool = Executors.newFixedThreadPool(10)

  def init() {
    credentials = new PropertiesCredentials(Thread.currentThread.getContextClassLoader.getResourceAsStream("AwsCredentials.properties"))
    client = new AmazonGlacierClient(credentials)
    client.setEndpoint("https://glacier.us-east-1.amazonaws.com/")
    atm = new ArchiveTransferManager(client, credentials)
  }

  def stop() {
    client.shutdown()
    pool.shutdownNow()
  }

  def upload(archive: File, vaultName: String, percent: Int) {
    if (!archive.exists()) {
      log.warn("Not uploading " + archive.getAbsolutePath + " because it doesn't exist")
      return
    }
    try {
      var bytesTransferred = 0L
      val totalBytes = FileUtils.getSizeInBytes(archive.getAbsolutePath)
      output.publish(PartialUpload(archive.getAbsolutePath, 0))
      val result: UploadResult = atm.upload("-", vaultName, archive.toString, archive, new ProgressListener {
        def progressChanged(progressEvent: ProgressEvent) {
          if (progressEvent.getEventType.isByteCountEvent) {
            if (progressEvent.getEventType.equals(ProgressEventType.REQUEST_CONTENT_LENGTH_EVENT)) {

            }
            bytesTransferred += progressEvent.getBytesTransferred
          }
          if (totalBytes > 0) {
            val partial: PartialUpload = PartialUpload(archive.getAbsolutePath, scala.math.round(100.0 * bytesTransferred / totalBytes).toInt)
            output.publish(partial)
          }
        }
      })
      log.info(String.format("%s|%s", archive.getAbsolutePath, result.getArchiveId))
      output.publish(CompleteUpload(archive.getPath, vaultName, result.getArchiveId, percent))
     } catch {
      case (e: Throwable) => output.publish(FailedUpload(archive.getPath, vaultName, e, percent))
    }
  }
}


//  def multipartUpload(archive: File, vaultName: String, percent: Int): Unit = {
//    val partSize = 1024 * 1024 * 2
//    val initiation = new InitiateMultipartUploadRequest(vaultName, "", partSize.toString)
//    val initResult = client.initiateMultipartUpload(initiation)
//
//    val filePosition = 0
//    var currentPosition = 0L
//    val buffer = Array.ofDim[Byte](partSize)
//
//    val fis = new FileInputStream(archive)
//    var contentRange: String = ""
//    var read = 0
//    var done = false
//
//    var records = List.empty[MultipartRecord]
//
//    while (!done && currentPosition < archive.length()) {
//      read = fis.read(buffer, filePosition, buffer.size)
//      if (read > -1) {
//        val bytesRead = util.Arrays.copyOf(buffer, read)
//        log.info("read %s bytes from %s".format(read, archive.toString))
//        contentRange = "bytes %s-%s/*".format(currentPosition, currentPosition + read - 1)
//        records = MultipartRecord(bytesRead, contentRange, initResult.getUploadId, vaultName) :: records
//        currentPosition = currentPosition + read
//      } else {
//        done = true
//      }
//    }
//
//    fis.close()
//
//    val callables: List[Callable[(String, Array[Byte])]] = records.map {
//      r => {
//        val c = new Callable[(String, Array[Byte])] {
//          override def call(): (String, Array[Byte]) = {
//            val checksum = TreeHashGenerator.calculateTreeHash(new ByteArrayInputStream(r.bytes))
//            val binaryChecksum = BinaryUtils.fromHex(checksum)
//            var partRequest = new UploadMultipartPartRequest
//            partRequest = partRequest.withVaultName(vaultName)
//            partRequest = partRequest.withBody(new ByteArrayInputStream(r.bytes))
//            partRequest = partRequest.withChecksum(checksum)
//            partRequest = partRequest.withRange(contentRange)
//            partRequest = partRequest.withUploadId(initResult.getUploadId)
//            client.uploadMultipartPart(partRequest)
//            (contentRange, binaryChecksum)
//          }
//        }
//        c
//      }
//    }
//    val futures = pool.invokeAll(callables)
//    val checksums = futures.map {f => f.get()}.sortBy(_._1).map(_._2)
//    val hash = TreeHashGenerator.calculateTreeHash(checksums)
//    val archiveId = completeMultipartUpload(initResult.getUploadId, hash, archive, vaultName)
//    output.publish(CompleteUpload(archive.getPath, vaultName, archiveId, percent))
//  }
//
//  def completeMultipartUpload(uploadId: String, checksum: String, archive: File, vaultName: String) = {
//    val request = new CompleteMultipartUploadRequest(vaultName, uploadId, archive.length.toString,checksum)
//    val result = client.completeMultipartUpload(request)
//    result.getLocation
//  }
//
//}

File: src/main/java/org/joshjoyce/lumivore/sync/GlacierUploader.java
package org.joshjoyce.lumivore.sync;

import com.amazonaws.auth.ClasspathPropertiesFileCredentialsProvider;
import com.amazonaws.client.builder.AwsClientBuilder;
import com.amazonaws.event.ProgressEvent;
import com.amazonaws.event.ProgressEventType;
import com.amazonaws.event.ProgressListener;
import com.amazonaws.regions.Regions;
import com.amazonaws.services.glacier.AmazonGlacier;
import com.amazonaws.services.glacier.AmazonGlacierClientBuilder;
import com.amazonaws.services.glacier.transfer.ArchiveTransferManager;
import com.amazonaws.services.glacier.transfer.ArchiveTransferManagerBuilder;
import org.apache.log4j.Logger;
import org.jetlang.channels.Channel;
import org.joshjoyce.lumivore.io.FileUtils;

import java.io.File;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class GlacierUploader {
    private static final Logger log = Logger.getLogger(GlacierUploader.class);

    private AmazonGlacier client;
    private ArchiveTransferManager atm;
    private final ExecutorService pool;
    private final Channel<UploadAttemptResult> output;

    private final String endpoint = "https://glacier.us-east-1.amazonaws.com/";
    private final Regions region = Regions.US_EAST_1;

    public GlacierUploader(Channel<UploadAttemptResult> output) {
        this.output = output;
        this.pool = Executors.newFixedThreadPool(10);
    }

    void init() {
        var builder =
                AmazonGlacierClientBuilder.standard()
                .withCredentials(new ClasspathPropertiesFileCredentialsProvider("AwsCredentials.properties"))
                .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(endpoint, region.getName()));

        client = builder.build();

        var atmBuilder = new ArchiveTransferManagerBuilder().withGlacierClient(client);
        atm = atmBuilder.build();
    }

    public void stop() {
        client.shutdown();
        pool.shutdownNow();
    }

    void upload(File archive, String vaultName, int percent) {
        if (!archive.exists()) {
            log.warn("Not uploading " + archive.getAbsolutePath() + " because it doesn't exist");
            return;
        }
        try {
            final var totalBytes = FileUtils.getSizeInBytes(archive.getAbsolutePath());
            output.publish(new UploadAttemptResult(UploadAttemptResult.Status.PartialUpload, archive.getAbsolutePath(), null, null, 0, null));

            var progressListener = new ProgressListener() {
                private long bytesTransferred;

                @Override
                public void progressChanged(ProgressEvent event) {
                    var type = event.getEventType();
                    if (type.isByteCountEvent()) {
                        if (type.equals(ProgressEventType.REQUEST_CONTENT_LENGTH_EVENT)) {
                            bytesTransferred += event.getBytesTransferred();
                        }
                    }
                    if (totalBytes > 0) {
                        int percent = (int) Math.round(100.0 * ((double) bytesTransferred) / totalBytes);
                        var partial = new UploadAttemptResult(UploadAttemptResult.Status.PartialUpload,
                            archive.getAbsolutePath(), null, null, percent, null);
                        output.publish(partial);
                    }
                }
            };
            var result = atm.upload("-", vaultName, archive.toString(), archive, progressListener);
            log.info(String.format("%s|%s", archive.getAbsolutePath(), result.getArchiveId()));
            var complete = new UploadAttemptResult(UploadAttemptResult.Status.Complete, archive.getPath(),
                    vaultName, result.getArchiveId(), percent, null);
            output.publish(complete);
        } catch (Exception e) {
            output.publish(new UploadAttemptResult(UploadAttemptResult.Status.FailedUpload, archive.getPath(),
                    vaultName, null, percent, e));
        }
    }
}
